pkg/csi/service/vanilla/controller.go

```diff
@@ -787,6 +788,11 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 	// Block Volume
 	node, err := c.nodeMgr.GetNodeByName(ctx, req.NodeId)
 	if err != nil {
+		// ControllerUnpublishVolume should succeed when the corresponding vm is already deleted.
```

Comment:
I have a concern. There are cases where the node manager has not yet synced the nodes(maybe the node events got delayed) and the ControllerUnpublish is invoked. In this case, we incorrect return success that the detach is successful.

Reply:
Observed both the times node manager update was immediate, after deleting vm from VC.

```
{"level":"info","time":"2021-01-29T18:24:33.9933385Z","caller":"node/manager.go:266","msg":"Successfully unregistered node with nodeName k8s-node-0397","TraceId":"1883f8f1-54ef-4722-bb45-81fc7821c67a"}
```

```
{"level":"info","time":"2021-01-29T18:59:00.966511893Z","caller":"node/manager.go:266","msg":"Successfully unregistered node with nodeName k8s-node-0491","TraceId":"090f7b3a-7e59-48c1-a169-45b134fb1d55"}
```

Also, we are querying the API server in GetNodeByName() if we anticipate any delay in the node events:
https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/common/cns-lib/node/manager.go#L139

Reply:
Here's the scenario. A pod is deleted when the CSI node is down. Now the CSI pod comes up, we now have a race between ControllerUnpublishVolume execution and node manager getting events for the current nodes in k8s. If ControllerUnpublishVolume runs first(before node manager is aware of any node), then they all succeed leaving behind the volumes attached to the node. The delete volume would never succeed since the volume is still attached. Do you agree?

Reply:
Handled this by invoking get nodes on k8s in node manager